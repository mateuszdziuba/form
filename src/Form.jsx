import React from 'react';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import FormControl from '@material-ui/core/FormControl';

import {
  validateEmpty,
  validateEmail,
  validatePassword,
  validateUsername,
} from './utils/validators';

const initialState = {
  yourName: {
    error: null,
    value: '',
    valid: undefined,
  },
  surname: {
    error: null,
    value: '',
    valid: undefined,
  },
  username: {
    error: null,
    value: '',
    valid: undefined,
  },
  email: {
    error: null,
    value: '',
    valid: undefined,
  },
  password: {
    error: null,
    value: '',
    valid: undefined,
  },
};

const buttonStyle = {
  minWidth: '230px',
  width: '50%',
  borderRadius: '24px',
  fontFamily: 'Roboto',
  fontSize: '22px',
  fontWeight: '500',
  lineHeight: '30px',
  padding: '15px 30px',
  marginTop: '15px',
  alignSelf: 'center',
  // backgroundImage: 'linear-gradient(to right, #3b92ff 0%, #0770f3 100%)',
};
const formStyle = {
  width: '50%',
  maxWidth: '900px',
  border: '1px solid #eef3f6',
  padding: '50px 80px',
  marginTop: '20px',
  fontSize: '15px',
  boxShadow: '0 2px 4px 0 rgba(0,0,0,0.16),0 2px 10px 0 rgba(0,0,0,0.12)'
};

class Form extends React.Component {
  constructor() {
    super();
    this.state = initialState;
  }

  handleChange = (e, validators) => {
    const { name, value } = e.target;
    let valid;
    let error;
    for (let i = 0; i < validators.length; i += 1) {
      const rule = validators[i];
      ({ valid, error } = rule(value));

      if (!valid) {
        break;
      }
    }
    this.props.onChange({ [name]: value });
    this.setState(prevState => ({
      [name]: {
        ...prevState[name],
        value,
        error,
        valid,
      },
    }));
  };

  validateField = (e, validators) => {
    const { name, value } = e.target;
    let valid;
    let error;
    for (let i = 0; i < validators.length; i += 1) {
      const rule = validators[i];
      ({ valid, error } = rule(value));

      if (!valid) {
        break;
      }
    }
    this.setState(prevState => ({
      [name]: {
        ...prevState[name],
        error,
        valid,
      },
    }));
  };

  onSubmit = (e) => {
    e.preventDefault();
    let valid;
    let error;
    Object.keys(this.state).forEach((field) => {
      if (this.state[field].error === null) {
        ({ valid, error } = validateEmpty(this.state[field].value));
        this.setState(prevState => ({
          [field]: {
            ...prevState[field],
            error,
            valid,
          },
        }));
      }
    });
    const {
      yourName, surname, username, email, password,
    } = this.state;
    if (yourName.valid && surname.valid && username.valid && email.valid && password.valid) {
      alert('Registered succesfully');
      this.setState(
        initialState,
      );
      const fields = {
        yourName: '', surname: '', username: '', email: '', password: '',
      };
      this.props.onChange(fields);
    }
  };

  render() {
    const {
      yourName,
      surname,
      username,
      email,
      password,
    } = this.state;
    return (
      <FormControl style={formStyle}>
        <h1>Sign up</h1>
        <TextField
          name="yourName"
          label="Your Name"
          value={yourName.value}
          onChange={e => this.handleChange(e, [validateEmpty])}
          error={yourName.valid === false}
          helperText={yourName.valid !== false ? ' ' : yourName.error}
          onBlur={e => this.validateField(e, [validateEmpty])}
          fullWidth
        />
        <TextField
          name="surname"
          label="Surname"
          value={surname.value}
          onChange={e => this.handleChange(e, [validateEmpty])}
          error={surname.valid === false}
          helperText={surname.valid !== false ? ' ' : surname.error}
          onBlur={e => this.validateField(e, [validateEmpty])}
          fullWidth
        />
        <TextField
          name="username"
          label="Username"
          value={username.value}
          onChange={e => this.handleChange(e, [validateEmpty, validateUsername])}
          error={username.valid === false}
          helperText={username.valid !== false ? ' ' : username.error}
          onBlur={e => this.validateField(e, [validateEmpty, validateUsername])}
          fullWidth
        />
        <TextField
          name="email"
          label="E-mail address"
          type="email"
          value={email.value}
          onChange={e => this.handleChange(e, [validateEmpty, validateEmail])}
          error={email.valid === false}
          helperText={email.valid !== false ? ' ' : email.error}
          onBlur={e => this.validateField(e, [validateEmpty, validateEmail])}
          fullWidth
        />
        <TextField
          name="password"
          label="Password"
          type="password"
          value={password.value}
          onChange={e => this.handleChange(e, [validateEmpty, validatePassword])}
          error={password.valid === false}
          helperText={password.valid !== false ? ' ' : password.error}
          onBlur={e => this.validateField(e, [validateEmpty, validatePassword])}
          fullWidth
        />
        <Button
          onClick={e => this.onSubmit(e)}
          variant="raised"
          color="primary"
          // size="large"
          style={buttonStyle}
          children="submit"
        />
      </FormControl>
    );
  }
}

export default Form;
