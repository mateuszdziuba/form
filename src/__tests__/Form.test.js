import { shallow } from 'enzyme';
import React from 'react';
import Form from '../Form';

const onChange = jest.fn();
const wrapper = shallow(<Form onChange={onChange} />);

describe('Form:', () => {
  it('expect to render Form component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('expect to render 5 TextFields and 1 Button', () => {
    expect(wrapper.find('TextField').length).toBe(5);
    expect(wrapper.find('WithStyles(Button)').exists()).toBe(true);
  });
});

describe('Events:', () => {
  it('should validate on change', () => {
    const event = {
      target: {
        name: 'yourName',
        value: 'Matti',
      },
    };
    const fields = {
      error: '',
      value: 'Matti',
      valid: true,
    };
    wrapper.find('#yourName').simulate('change', event);
    expect(wrapper.state('yourName')).toEqual(fields);
  });

  it('should show error on change', () => {
    const event = {
      target: {
        name: 'username',
        value: 'Matt',
      },
    };
    const fields = {
      error: 'Username must be at least 5 characters long',
      value: 'Matt',
      valid: false,
    };
    wrapper.find('#username').simulate('change', event);
    expect(wrapper.state('username')).toEqual(fields);
  });

  it('should validate on blur', () => {
    const event = {
      target: {
        name: 'surname',
        value: '',
      },
    };
    const fields = {
      error: 'This field is required',
      value: '',
      valid: false,
    };
    wrapper.find('#surname').simulate('blur', event);
    expect(wrapper.state('surname')).toEqual(fields);
  });

  it('should validate on submit', () => {
    const state = {
      yourName: {
        error: '',
        value: 'Matti',
        valid: true,
      },
      surname: {
        error: 'This field is required',
        value: '',
        valid: false,
      },
      username: {
        error: 'Username must be at least 5 characters long',
        value: 'Matt',
        valid: false,
      },
      email: {
        error: 'This field is required',
        value: '',
        valid: false,
      },
      password: {
        error: 'This field is required',
        value: '',
        valid: false,
      },
    };
    wrapper.find('#button').simulate('click');
    expect(wrapper.state()).toEqual(state);
  });
});
