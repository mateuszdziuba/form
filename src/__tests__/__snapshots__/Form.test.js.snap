// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form: expect to render Form component 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Form
    onChange={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <h1>
          Sign up
</h1>,
        <TextField
          error={false}
          fullWidth={true}
          helperText=" "
          id="yourName"
          label="Your Name"
          name="yourName"
          onBlur={[Function]}
          onChange={[Function]}
          required={false}
          select={false}
          value=""
/>,
        <TextField
          error={false}
          fullWidth={true}
          helperText=" "
          id="surname"
          label="Surname"
          name="surname"
          onBlur={[Function]}
          onChange={[Function]}
          required={false}
          select={false}
          value=""
/>,
        <TextField
          error={false}
          fullWidth={true}
          helperText=" "
          id="username"
          label="Username"
          name="username"
          onBlur={[Function]}
          onChange={[Function]}
          required={false}
          select={false}
          value=""
/>,
        <TextField
          error={false}
          fullWidth={true}
          helperText=" "
          id="email"
          label="E-mail address"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          required={false}
          select={false}
          type="email"
          value=""
/>,
        <TextField
          error={false}
          fullWidth={true}
          helperText=" "
          id="password"
          label="Password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          required={false}
          select={false}
          type="password"
          value=""
/>,
        <WithStyles(Button)
          color="primary"
          id="button"
          onClick={[Function]}
          style={
                    Object {
                              "alignSelf": "center",
                              "borderRadius": "24px",
                              "fontFamily": "Roboto",
                              "fontSize": "22px",
                              "fontWeight": "500",
                              "lineHeight": "30px",
                              "marginTop": "15px",
                              "minWidth": "230px",
                              "padding": "15px 30px",
                              "width": "50%",
                            }
          }
          variant="raised"
>
          submit
</WithStyles(Button)>,
      ],
      "style": Object {
        "border": "1px solid #eef3f6",
        "boxShadow": "0 2px 4px 0 rgba(0,0,0,0.16),0 2px 10px 0 rgba(0,0,0,0.12)",
        "fontSize": "15px",
        "marginTop": "20px",
        "maxWidth": "900px",
        "padding": "50px 80px",
        "width": "50%",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Sign up",
        },
        "ref": null,
        "rendered": "Sign up",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "error": false,
          "fullWidth": true,
          "helperText": " ",
          "id": "yourName",
          "label": "Your Name",
          "name": "yourName",
          "onBlur": [Function],
          "onChange": [Function],
          "required": false,
          "select": false,
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "error": false,
          "fullWidth": true,
          "helperText": " ",
          "id": "surname",
          "label": "Surname",
          "name": "surname",
          "onBlur": [Function],
          "onChange": [Function],
          "required": false,
          "select": false,
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "error": false,
          "fullWidth": true,
          "helperText": " ",
          "id": "username",
          "label": "Username",
          "name": "username",
          "onBlur": [Function],
          "onChange": [Function],
          "required": false,
          "select": false,
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "error": false,
          "fullWidth": true,
          "helperText": " ",
          "id": "email",
          "label": "E-mail address",
          "name": "email",
          "onBlur": [Function],
          "onChange": [Function],
          "required": false,
          "select": false,
          "type": "email",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "error": false,
          "fullWidth": true,
          "helperText": " ",
          "id": "password",
          "label": "Password",
          "name": "password",
          "onBlur": [Function],
          "onChange": [Function],
          "required": false,
          "select": false,
          "type": "password",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "submit",
          "color": "primary",
          "id": "button",
          "onClick": [Function],
          "style": Object {
            "alignSelf": "center",
            "borderRadius": "24px",
            "fontFamily": "Roboto",
            "fontSize": "22px",
            "fontWeight": "500",
            "lineHeight": "30px",
            "marginTop": "15px",
            "minWidth": "230px",
            "padding": "15px 30px",
            "width": "50%",
          },
          "variant": "raised",
        },
        "ref": null,
        "rendered": "submit",
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <h1>
            Sign up
</h1>,
          <TextField
            error={false}
            fullWidth={true}
            helperText=" "
            id="yourName"
            label="Your Name"
            name="yourName"
            onBlur={[Function]}
            onChange={[Function]}
            required={false}
            select={false}
            value=""
/>,
          <TextField
            error={false}
            fullWidth={true}
            helperText=" "
            id="surname"
            label="Surname"
            name="surname"
            onBlur={[Function]}
            onChange={[Function]}
            required={false}
            select={false}
            value=""
/>,
          <TextField
            error={false}
            fullWidth={true}
            helperText=" "
            id="username"
            label="Username"
            name="username"
            onBlur={[Function]}
            onChange={[Function]}
            required={false}
            select={false}
            value=""
/>,
          <TextField
            error={false}
            fullWidth={true}
            helperText=" "
            id="email"
            label="E-mail address"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            required={false}
            select={false}
            type="email"
            value=""
/>,
          <TextField
            error={false}
            fullWidth={true}
            helperText=" "
            id="password"
            label="Password"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            required={false}
            select={false}
            type="password"
            value=""
/>,
          <WithStyles(Button)
            color="primary"
            id="button"
            onClick={[Function]}
            style={
                        Object {
                                    "alignSelf": "center",
                                    "borderRadius": "24px",
                                    "fontFamily": "Roboto",
                                    "fontSize": "22px",
                                    "fontWeight": "500",
                                    "lineHeight": "30px",
                                    "marginTop": "15px",
                                    "minWidth": "230px",
                                    "padding": "15px 30px",
                                    "width": "50%",
                                  }
            }
            variant="raised"
>
            submit
</WithStyles(Button)>,
        ],
        "style": Object {
          "border": "1px solid #eef3f6",
          "boxShadow": "0 2px 4px 0 rgba(0,0,0,0.16),0 2px 10px 0 rgba(0,0,0,0.12)",
          "fontSize": "15px",
          "marginTop": "20px",
          "maxWidth": "900px",
          "padding": "50px 80px",
          "width": "50%",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Sign up",
          },
          "ref": null,
          "rendered": "Sign up",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "error": false,
            "fullWidth": true,
            "helperText": " ",
            "id": "yourName",
            "label": "Your Name",
            "name": "yourName",
            "onBlur": [Function],
            "onChange": [Function],
            "required": false,
            "select": false,
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "error": false,
            "fullWidth": true,
            "helperText": " ",
            "id": "surname",
            "label": "Surname",
            "name": "surname",
            "onBlur": [Function],
            "onChange": [Function],
            "required": false,
            "select": false,
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "error": false,
            "fullWidth": true,
            "helperText": " ",
            "id": "username",
            "label": "Username",
            "name": "username",
            "onBlur": [Function],
            "onChange": [Function],
            "required": false,
            "select": false,
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "error": false,
            "fullWidth": true,
            "helperText": " ",
            "id": "email",
            "label": "E-mail address",
            "name": "email",
            "onBlur": [Function],
            "onChange": [Function],
            "required": false,
            "select": false,
            "type": "email",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "error": false,
            "fullWidth": true,
            "helperText": " ",
            "id": "password",
            "label": "Password",
            "name": "password",
            "onBlur": [Function],
            "onChange": [Function],
            "required": false,
            "select": false,
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "submit",
            "color": "primary",
            "id": "button",
            "onClick": [Function],
            "style": Object {
              "alignSelf": "center",
              "borderRadius": "24px",
              "fontFamily": "Roboto",
              "fontSize": "22px",
              "fontWeight": "500",
              "lineHeight": "30px",
              "marginTop": "15px",
              "minWidth": "230px",
              "padding": "15px 30px",
              "width": "50%",
            },
            "variant": "raised",
          },
          "ref": null,
          "rendered": "submit",
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
